common/src/main/java/io/netty/util/Recycler.java

```diff
@@ -176,486 +174,120 @@ public final boolean recycle(T o, Handle<T> handle) {
-                    // We never unlink the first queue, as we don't want to synchronize on updating the head.
-                    if (cursor.hasFinalData()) {
-                        for (;;) {
-                            if (cursor.transfer(this)) {
-                                success = true;
-                            } else {
-                                break;
-                            }
-                        }
-                    }
+                int maxId;
+                if (i < genMaxId.length) {
+                    maxId = genMaxId[i];
+                } else {
+                    // Eden generation
 
-                    if (prev != null) {
-                        prev.setNext(next);
+                    maxId = parent.currentGenId.get();
+                    if (maxId == genMaxId[i - 1]) {
+                        // This is the first thread to detect the new generation
+                        parent.currentGenId.incrementAndGet();
                     }
-                } else {
-                    prev = cursor;
                 }
-
-                cursor = next;
-
-            } while (cursor != null && !success);
-
-            this.prev = prev;
-            this.cursor = cursor;
-            return success;
-        }
-
-        void push(DefaultHandle<?> item) {
-            Thread currentThread = Thread.currentThread();
-            if (threadRef.get() == currentThread) {
-                // The current Thread is the thread that belongs to the Stack, we can try to push the object now.
-                pushNow(item);
-            } else {
-                // The current Thread is not the one that belongs to the Stack
-                // (or the Thread that belonged to the Stack was collected already), we need to signal that the push
-                // happens later.
-                pushLater(item, currentThread);
+                genMaxId[i - 1] = maxId;
             }
         }
 
-        private void pushNow(DefaultHandle<?> item) {
-            if ((item.recycleId | item.lastRecycledId) != 0) {
+        void push(DefaultHandle<T> node) {
+            if (node.next != null) {
                 throw new IllegalStateException("recycled already");
             }
-            item.recycleId = item.lastRecycledId = OWN_THREAD_ID;
-
-            int size = this.size;
-            if (size >= maxCapacity || dropHandle(item)) {
-                // Hit the maximum capacity or should drop - drop the possibly youngest object.
-                return;
-            }
-            if (size == elements.length) {
-                elements = Arrays.copyOf(elements, min(size << 1, maxCapacity));
+            int i = 0;
+            for (; i < genMaxId.length && genMaxId[i] < node.genId; i++) {
+                // find generation index
             }
-
-            elements[size] = item;
-            this.size = size + 1;
+            node.next = genHead[i];
+            genHead[i] = node.selfRef;
```

Comment:
The array business is not obvious to me. Here we are relying on `genMaxId.length` being one less than `genHead.length`, which feels shaky.

Reply:
They are parallel arrays and need to have that relationship; they are _very_ closely related. We could call it out more in the documentation, if that would help.

Reply:
The `genMaxId` field javadoc comment could be expanded to explain how these two arrays work together. I think that would be useful as this code is a little unusual :)